plugin/dnstap/io.go

```diff
@@ -24,31 +26,39 @@ type tapper interface {
 	Dnstap(*tap.Dnstap)
 }
 
+type WarnLogger interface {
+	Warningf(format string, v ...interface{})
+}
+
 // dio implements the Tapper interface.
 type dio struct {
-	endpoint        string
-	proto           string
-	enc             *encoder
-	queue           chan *tap.Dnstap
-	dropped         uint32
-	quit            chan struct{}
-	flushTimeout    time.Duration
-	tcpTimeout      time.Duration
-	skipVerify      bool
-	tcpWriteBufSize int
+	endpoint          string
+	proto             string
+	enc               *encoder
+	queue             chan *tap.Dnstap
+	dropped           uint32
+	quit              chan struct{}
+	flushTimeout      time.Duration
+	errorCheckTimeout time.Duration
+	tcpTimeout        time.Duration
+	skipVerify        bool
+	tcpWriteBufSize   int
+	logger            WarnLogger
 }
 
 // newIO returns a new and initialized pointer to a dio.
 func newIO(proto, endpoint string, multipleQueue int, multipleTcpWriteBuf int) *dio {
 	return &dio{
-		endpoint:        endpoint,
-		proto:           proto,
-		queue:           make(chan *tap.Dnstap, multipleQueue*queueSize),
-		quit:            make(chan struct{}),
-		flushTimeout:    flushTimeout,
-		tcpTimeout:      tcpTimeout,
-		skipVerify:      skipVerify,
-		tcpWriteBufSize: multipleTcpWriteBuf * tcpWriteBufSize,
+		endpoint:          endpoint,
+		proto:             proto,
+		queue:             make(chan *tap.Dnstap, multipleQueue*queueSize),
+		quit:              make(chan struct{}),
+		flushTimeout:      flushTimeout,
+		errorCheckTimeout: errorCheckTimeout,
```

Comment:
if you are not going to make it configurable, you don't need this property in `dio`, just use the constant where needed